### ISSUE TITLE
segfault when assigning array.size to Py_ssize_t variable with alpha release

### ISSUE THREAD

**Issue by emmanuelle** [cython/cython#3520]
Hi Cython developers, thank you for publishing an alpha release, this is very useful for libraries such as scikit-image. One of our tests segfaults with the alpha, and we tracked it down to a line doing
```
cdef Py_ssize_t a = arr.size
```
where `arr` is here a 1-D numpy array of floats (`cdef cnp.ndarray[cnp.float_t, ndim=1] arr`). When we change this line to
```
cdef Py_ssize_t a = len(arr)
```
the segfault disappears and the function behaves as expected. The problem did not exist with cython 0.29.16. 

So the question for you is: is this change wanted and should we upgrade our code? Or is it a bug on the Cython end? Thank you in advance :-).

The corresponding issue in scikit-image is https://github.com/scikit-image/scikit-image/issues/4577

**Comment by scoder** [comment]
Thanks for the report @emmanuelle. The difference in Cython 3.0 is that this property no longer accesses the NumPy internals directly but calls `PyArray_SIZE()` instead. This should … generally … work.

In any case, all segfaults from Cython generated code are bugs. According to the ticket you linked, the code goes more or less like this:
```
    cdef cnp.ndarray[cnp.float_t, ndim=1] costs = np.hstack([
    	right_cost.ravel(), down_cost.ravel(), dright_cost.ravel(),
        uright_cost.ravel()]).astype(np.float)

    edge_queue = np.argsort(costs)
    costs = np.ascontiguousarray(costs[edge_queue])

    cdef Py_ssize_t num_costs = costs.size
```
And the last line then crashes in (the expected)
```
4992	  __pyx_t_45 = PyArray_SIZE(__pyx_v_costs);
```
@mattip any idea what might be going wrong here?

**Comment by da-woods** [comment]
I'm wondering if it's just missing a call to `cnp.import_array()`?

**Comment by mattip** [comment]
Whoops. Probably because the definition is wrong :(. [It is](https://github.com/cython/cython/blob/3.0a1/Cython/Includes/numpy/__init__.pxd#L265) `return PyArray_SIZE(ndarray)` not `return PyArray_SIZE(self)`.

I will add a test and fix

**Comment by scoder** [comment]
@mattip – yeah, that looks wrong. :)

But what @da-woods wrote might still apply, right? `PyArray_SIZE()` [is actually more complex](https://github.com/numpy/numpy/blob/24189be60ec4d823f4068be727b5d28f7eacc823/numpy/core/include/numpy/ndarrayobject.h#L73) than it seems.

**Comment by mattip** [comment]
True. I wrote a quick test, it confirms without `import_array()` it segfaults, with `import_array()` it returns the correct answer, even with the seemingly incorrect `PyArray_SIZE(ndarray)`

**Comment by mattip** [comment]
xref gh-3521 (once CI finishes I will fix `numpy/__init__.pxd`)

**Comment by da-woods** [comment]
It looks like the changes to handling Numpy have made `np.import_array()` more necessary (in the past you could often get away without it, and now it looks like you can't).

Is it worth either generating a warning if it's missing or even special casing `cimport numpy` to do it automatically? Are there ever cases where you shouldn't do it?

I'm just thinking that there's quite a bit of code in existence without `import_array` that could suddenly "break".

**Comment by mattip** [comment]
> special casing `cimport numpy` to do it automatically

I guess it could be as easy as adding that to the `pxd`. Is there a precedent for a `pxd` being more than just a cython version of a header file and actually calling code?

**Comment by scoder** [comment]
> there's quite a bit of code in existence without import_array that could suddenly "break".

Yes, I was thinking that, too. We should find a way to make this safer.

> Is there a precedent for a pxd being more than just a cython version of a header file and actually calling code?

That's a bit the thing, you can't really (and shouldn't) execute code from a .pxd file. They are there to share declarations (and inline functions, as a special enough case to break the rules). They shouldn't have side effects. And it would also be difficult to know _when_ such a side-effect should occur exactly. Is it enough to do this at load time? At the start of the module init function? Do we need to allow users to do something before that? Should we do it at the point where the `cimport` is in the code? Would users expect that cimport to have a side-effect directly, even if they then later use a Python import for the same module as well? Difficult.

I'm not saying it can't be done, just that it's not something that's currently there, and that it seems worth trying to search for other ways before we add code execution to cimports.

**Comment by scoder** [comment]
Seeing the good ideas in #3524 made me take a step back. What do you think about this:

There is no actual code currently being generated for cimports (in `CImportStatNode` and `FromCImportStatNode`), but we have a mechanism for additional loading utility code for certain modules, so there is a precedent for special-casing certain modules.

IF `numpy` has been cimported in the current module, AND neither `import_array` nor `_import_array` have not been used anywhere (either by calling it or by writing `(void) import_array` and not calling it), THEN we insert a call to `_import_array()` in the place of the cimport automatically.

That way, most users won't have to care any more, and users who do care have a way to explicitly call the function in the right place or to prevent it from being called entirely.

There's also support in the utility code handling mechanism (`"MyUtilCodeSection.init"`) for inserting code into the module init function (which I would prefer since it uses the existing mechanisms), but I'm not sure if that inserts the code in a suitable place. To be tried, I guess.

**Comment by da-woods** [comment]
I think that sounds reasonable. I don't _really_ like the special-casing, but Numpy is clearly special-cased all over Cython so it's not adding anything new. People tend not to read warnings anyway (and these are buried in quite a bit of compiler output) so I think just fixing the problem is sensible.

The init utility code looks to insert into the right place. I've updated the PR to do this.

For the moment I'd think it might be best not to update the documentation - keep it showing explicit calls to `np.import_array()` and keep this feature semi-undocumented? That was if there's any cases that it misses (multiple modules compiled at once, pxds cimported from pxds, etc?) it's still a user bug not a Cython bug?

**Comment by scoder** [comment]
@da-woods I'd give it at least some in-the-wild testing before we consider it an official feature.

**Comment by scoder** [comment]
Fixed in #3524.
